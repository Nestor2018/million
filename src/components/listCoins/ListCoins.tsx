import React, { useState, useEffect, useCallback } from 'react';
import { SafeAreaView, FlatList, StyleSheet, Text } from 'react-native';
import { NavigationProp } from '@react-navigation/native';

import CoinItem from './CoinItem';
import CoinSearch from '../coinSearch/CoinSearch';
import Loader from '../loader/Loader';
import { get } from '../../api/coins';
import colors from '../../utils/colors';
import { Icoin } from '../../interfaces/coin.interface';


interface ListCoinsProps {
  navigation: NavigationProp<any>;
}

/**
 * Componente ListCoins (lista de monedas)
 * @param {ListCoinsProps} props - Propiedades que contienen el objeto de navegación
 * @returns {JSX.Element} El componente renderizado
 */
const ListCoins: React.FC<ListCoinsProps> = ({ navigation }) => {
  const [coins, setCoins] = useState<Icoin[]>([]);
  const [allCoins, setAllCoins] = useState<Icoin[]>([]);
  const [loading, setLoading] = useState<boolean>(false);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    (async () => {
      await loadCoins();
    })();
  }, []);

  /**
  Realiza una solicitud GET para obtener las monedas
  */

  const loadCoins = async () => {
    try {
      setLoading(true);
      const response = await get('/tickers');
      setCoins(response.data);
      setAllCoins(response.data);
    } catch (error) {
      console.error('Error loading coins:', error);
      setError('Error loading coins. Please try again later.');
    } finally {
      setLoading(false);
    }
  };

  /**
  Recibe el texto ingresado en el input para filtrar el array de monedas
  @param {string} query
  */
  const handleSearch = useCallback((query: string) => {
    const coinsFiltered = allCoins.filter((coin: Icoin) => {
      return (
        coin.name.toLowerCase().includes(query.toLowerCase()) ||
        coin.symbol.toLowerCase().includes(query.toLowerCase())
      );
    });
    setCoins(coinsFiltered);
  }, [allCoins]);

  /**
  Redirige a la pantalla de CoinDetail y envía la moneda para ser renderizada
  @param {object} coin 
  */
  const handlePress = useCallback((coin: Icoin) => {
    navigation.navigate('CoinDetail', { coin });
  }, [navigation]);

  const renderItem = ({ item }: { item: Icoin }) => (
    <CoinItem coin={item} onPress={() => handlePress(item)} />
  );

  return (
    <SafeAreaView style={styles.container} testID="listCoins">
      <CoinSearch onChange={handleSearch} />
      {loading ? <Loader testID="loader" /> : null}
      {error ? <Text style={styles.errorText}>{error}</Text> : null}
      <FlatList
        data={coins}
        renderItem={renderItem}
        keyExtractor={item => item.id}
        extraData={allCoins}
        initialNumToRender={13}
      />
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    backgroundColor: colors.charade,
    flex: 1,
  },
  errorText: {
    color: 'red',
    textAlign: 'center',
    marginVertical: 10,
  },
});

export default ListCoins;

