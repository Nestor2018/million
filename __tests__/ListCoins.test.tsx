import React from 'react';
import { render, fireEvent, act } from '@testing-library/react-native';
import ListCoins from '../src/components/listCoins/ListCoins';
import * as coinsApi from '../src/api/coins';

const mockCoins = [
  {
    id: '1',
    name: 'Bitcoin',
    symbol: 'BTC',
    price_usd: '50000',
    percent_change_1h: 5,
  },
  {
    id: '2',
    name: 'Ethereum',
    symbol: 'ETH',
    price_usd: '3000',
    percent_change_1h: -2,
  },
];

// Mock del navigation props
const mockNavigation = { navigate: jest.fn() };

describe('ListCoins Component', () => {
  beforeEach(() => {
    jest.spyOn(coinsApi, 'get').mockResolvedValue({ data: mockCoins });
  });

  it('should filter coins correctly', async () => {
    const { getByTestId, queryByText } = render(
      <ListCoins navigation={mockNavigation} />
    );

    // Espera a que los datos se carguen y el componente se actualice
    await act(async () => {
      await Promise.resolve();
    });

    // Filtra las monedas por nombre
    const coinSearchInput = getByTestId('input');
    act(() => {
      fireEvent.changeText(coinSearchInput, 'Bitcoin');
    });

    // Solo Bitcoin debería ser visible
    expect(queryByText('Bitcoin')).toBeTruthy();
    expect(queryByText('Ethereum')).toBeNull();
  });

  it('should load coins from the API on mount', async () => {
    const { getByTestId } = render(<ListCoins navigation={mockNavigation} />);

    // Espera a que los datos se carguen y el componente se actualice
    await act(async () => {
      await Promise.resolve();
    });

    expect(getByTestId('listCoins')).toBeTruthy();
  });

  it('should show loader while loading data', async () => {
    const { getByTestId } = render(<ListCoins navigation={mockNavigation} />);

    // Verifica que el Loader esté presente mientras los datos se cargan
    expect(getByTestId('loader')).toBeTruthy();

    // Espera a que los datos se carguen
    await act(async () => {
      await Promise.resolve();
    });

    // Verifica que el Loader desaparezca después de cargar los datos
    expect(() => getByTestId('loader')).toThrow();
  });

  it('should show all coins when search input is cleared', async () => {
    const { getByTestId, queryByText } = render(
      <ListCoins navigation={mockNavigation} />
    );

    // Espera a que los datos se carguen
    await act(async () => {
      await Promise.resolve();
    });

    // Filtra las monedas por nombre
    const coinSearchInput = getByTestId('input');
    fireEvent.changeText(coinSearchInput, 'Bitcoin');

    // Solo Bitcoin debería ser visible
    expect(queryByText('Bitcoin')).toBeTruthy();
    expect(queryByText('Ethereum')).toBeNull();

    // Limpia el campo de búsqueda
    fireEvent.changeText(coinSearchInput, '');

    // Todas las monedas deberían ser visibles nuevamente
    expect(queryByText('Bitcoin')).toBeTruthy();
    expect(queryByText('Ethereum')).toBeTruthy();
  });

  it('should handle edge cases in search input', async () => {
    const { getByTestId, queryByText } = render(
      <ListCoins navigation={mockNavigation} />
    );

    // Espera a que los datos se carguen
    await act(async () => {
      await Promise.resolve();
    });

    // Simula una búsqueda con texto largo
    const coinSearchInput = getByTestId('input');
    fireEvent.changeText(coinSearchInput, 'a'.repeat(50));

    // Ninguna moneda debería coincidir
    expect(queryByText('Bitcoin')).toBeNull();
    expect(queryByText('Ethereum')).toBeNull();

    // Simula una búsqueda con caracteres especiales
    fireEvent.changeText(coinSearchInput, '@#$%^&*');

    // Ninguna moneda debería coincidir
    expect(queryByText('Bitcoin')).toBeNull();
    expect(queryByText('Ethereum')).toBeNull();
  });

  it('matches snapshot', async () => {
    const tree = render(<ListCoins navigation={mockNavigation} />);

    // Espera a que los datos se carguen
    await act(async () => {
      await Promise.resolve();
    });

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should render initial number of items correctly', async () => {
    const { getAllByTestId } = render(<ListCoins navigation={mockNavigation} />);

    // Espera a que los datos se carguen
    await act(async () => {
      await Promise.resolve();
    });

    // Verifica que se rendericen los elementos iniciales
    const renderedItems = getAllByTestId('coinItem');
    expect(renderedItems.length).toBe(2); // Deberían renderizarse 2 monedas (mockCoins)
  });
});
