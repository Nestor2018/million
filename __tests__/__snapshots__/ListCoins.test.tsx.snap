// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListCoins Component matches snapshot 1`] = `
<RCTSafeAreaView
  testID="listCoins"
>
  <View
    testID="coinSearch"
  >
    <TextInput
      accessibilityLabel="Search coin"
      accessible={true}
      onChangeText={[Function]}
      placeholder="Search coin"
      placeholderTextColor="white"
      style={
        [
          undefined,
          undefined,
        ]
      }
      testID="input"
      value=""
    />
  </View>
  <View
    testID="flat-list"
  >
    <View
      testID="flat-list-item-0"
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="coinItem"
      >
        <View>
          <Text>
            BTC
          </Text>
          <Text>
            Bitcoin
          </Text>
          <Text>
             
            $50000
          </Text>
        </View>
        <View>
          <Text>
            5
          </Text>
          <Image
            source={
              {
                "testUri": "../../../assets/arrow_up.png",
              }
            }
          />
        </View>
      </View>
    </View>
    <View
      testID="flat-list-item-1"
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="coinItem"
      >
        <View>
          <Text>
            ETH
          </Text>
          <Text>
            Ethereum
          </Text>
          <Text>
             
            $3000
          </Text>
        </View>
        <View>
          <Text>
            -2
          </Text>
          <Image
            source={
              {
                "testUri": "../../../assets/arrow_down.png",
              }
            }
          />
        </View>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
